src/demo/containers/playground/FormServicesIntegrationDemo.tsx

```diff
@@ -0,0 +1,173 @@
+        .then((data: FormData) => {
+          console.log('Successfully updated FormData', data)
+          this.setState({ formData: data })
+        })
+        .catch((error: any) => console.error('Failed to update FormData', error))
+    }).catch((error: any) => console.error('Failed to update a new file handle:', error))
+  }
+
+  public onCreateFormData(): void {
+    SynapseClient.uploadFile(this.state.token, 'my test file', new Blob(['hello form data'])).then((fileUploadComplete: FileUploadComplete) => {
+      console.log('Successfully created test file handle, now create the form data!')
+      SynapseClient.createFormData(this.state.formGroup!.groupId!, 'my test form data', fileUploadComplete.fileHandleId, this.state.token!)
+        .then((data: FormData) => {
+          console.log('Successfully created FormData', data)
+          this.setState({ formData: data })
+        })
+        .catch((error: any) => console.error('Failed to create FormData', error))
+    }).catch((error: any) => console.error('Failed to create a new file handle:', error))
+  }
+
+  public onSubmitFormData(): void {
+    SynapseClient.submitFormData(this.state.formData!.formDataId!, this.state.token!)
+      .then((data: FormData) => {
+        console.log('Successfully submitted FormData', data)
+      })
+      .catch((error: any) => console.error('Failed to submit form data', error))
+  }
+
+  private listMyFormData = () => {
+    if (this.state.token && this.state.formGroup && !this.state.myFormDataList) {
+      let request: ListRequest = {
+        groupId: this.state.formGroup.groupId!,
+        filterByState: [StatusEnum.ACCEPTED, StatusEnum.SUBMITTED_WAITING_FOR_REVIEW, StatusEnum.WAITING_FOR_SUBMISSION]
+      }
+      SynapseClient.listFormData(request, this.state.token).then((resp: ListResponse) => {
+        console.log('Successfully got a my FormData list', resp)
+        this.setState({ myFormDataList: resp })
+      }).catch((error: any) => console.error('Failed to get my FormData list', error))
+    }
+  }
+
+  private listFormAdminFormData = () => {
+    if (this.state.token && this.state.formGroup && !this.state.formAdminFormDataList) {
+      let request: ListRequest = {
+        groupId: this.state.formGroup.groupId!,
+        filterByState: [StatusEnum.ACCEPTED, StatusEnum.SUBMITTED_WAITING_FOR_REVIEW]
+      }
+      SynapseClient.listFormDataAsFormAdmin(request, this.state.token).then((resp: ListResponse) => {
+        console.log('Successfully got FormGroup admin list', resp)
+        this.setState({ formAdminFormDataList: resp })
+      }).catch((error: any) => console.error('Failed to get FormGroup admin FormData list', error))
+    }
+  }
+
+  public componentDidUpdate() {
+    this.listMyFormData()
+    this.listFormAdminFormData()
+  }
+
+  render() {
+    return (
+      <TokenContext.Consumer>
+        {
+          (token: string) => {
+            if (token !== this.state.token) {
+              this.setState({
```

Comment:
noooooo

Reply:
could you elaborate on this point?

Reply:
Ideally setState wouldn't get called in render and instead in componentDidMount since setState is async and it kind of goes against the idea of render being pure.

Reply:
This makes sense, thanks Michael!